File: Sources/CountriesFlagsObjC/source/Countries/UIImage+PuertoRico.m
//
//  UIImage+PuertoRico.m
//  CountriesFlags
//

#import "Countries/UIImage+PuertoRico.h"

#import "Extensions/UIBezierPath+Star.h"
#import "Extensions/UIBezierPath+Triangle.h"

@implementation UIImage (PuertoRico)

+ (UIImage * _Nullable)flagPuertoRicoWithSize:(CGSize)size {
    UIColor* colorBlue = [UIColor colorWithRed:8.0f/255.0f
                                         green:68.0f/255.0f
                                          blue:1.0f
                                         alpha:1.0f];
    UIColor* colorRed = [UIColor colorWithRed:237.0f/255.0f
                                        green:0.0f
                                         blue:0.0f
                                        alpha:1.0f];
    UIColor* colorWhite = [UIColor whiteColor];
    
    UIGraphicsBeginImageContextWithOptions(size, NO, 0.0f);
    
    [colorRed setFill];
    UIRectFill(CGRectMake(0.0f, 0.0f, size.width, size.height));
    
    [colorWhite setFill];
    UIRectFill(CGRectMake(0.0f, size.height / 30.0f * 6.0f, size.width, size.height / 30.0f * 6.0f));
    UIRectFill(CGRectMake(0.0f, size.height / 30.0f * 18.0f, size.width, size.height / 30.0f * 6.0f));
    
    [colorBlue setFill];
    [[UIBezierPath triangleInSize:size type:FlagTriangleTypeLeftEquilateral] fill];
    
    [colorWhite setFill];
    [[UIBezierPath starWithSize:CGSizeMake(size.width / 45.0 * 8.660254037844386 * 2.0, size.height)
                         radius:size.height / 30.0f * 10.0f / 2.0f
                           type:FlagStarTypeFivePointed] fill];
    
    UIImage* image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return image;
}

@end

File: Sources/CountriesFlags/Countries/UIImage+PuertoRico.swift
//
//  UIImage+PuertoRico.swift
//  CountriesFlags
//

#if canImport(UIKit)

import UIKit

extension UIImage {
    /**
     [wikipedia]: https://en.wikipedia.org/wiki/Flag_of_Puerto_Rico "wikipedia"
     
     Generates the flag of Puerto Rico.
     For more information see [wikipedia].
     
     - Parameters:
        - size: image size.
     
     - Returns: Returns the flag of Puerto Rico.
     */
    internal static func flagPuertoRico(with size: CGSize) -> UIImage {
        let colorBlue = UIColor(red: 8.0 / 255.0,
                                green: 68.0 / 255.0,
                                blue: 1.0,
                                alpha: 1.0)
        let colorRed = UIColor(red: 237.0 / 255.0,
                               green: 0.0,
                               blue: 0.0,
                               alpha: 1.0)
        let colorWhite = UIColor.white

        UIGraphicsBeginImageContextWithOptions(size, true, 0.0)

        colorRed.setFill()
        UIRectFill(CGRect(origin: .zero, size: size))

        colorWhite.setFill()
        UIRectFill(CGRect(origin: CGPoint(x: 0, y: size.height / 30.0 * 6.0),
                          size: CGSize(width: size.width, height: size.height / 30.0 * 6.0)))
        UIRectFill(CGRect(origin: CGPoint(x: 0, y: size.height / 30.0 * 18.0),
                          size: CGSize(width: size.width, height: size.height / 30.0 * 6.0)))

        colorBlue.setFill()
        UIBezierPath.triangle(in: size, type: .leftEquilateral).fill()

        colorWhite.setFill()
        UIBezierPath.star(at: CGPoint(x: size.width / 45.0 * 8.660254037844386, y: size.height / 2.0),
                          radius: size.height / 30.0 * 10.0 / 2.0).fill()

        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()

        return image ?? UIImage()
    }
}

#endif
